### PR TITLE
feat: add gitea notifier

### PR DESCRIPTION
**Pull request by ttys3**
add support to send event notification to Gitea repo commit status

image

Gitea is a forge software package for hosting software development version control using Git as well as other collaborative features like bug tracking, code review, kanban boards, tickets, and wikis. It supports self-hosting but also provides a free public first-party instance

https://github.com/go-gitea/gitea

### CODE REVIEW

File: internal/notifier/gitea.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
+package notifier
+
+import (
```
Comment (makkes) [inline internal/notifier/gitea.go:3]: Please arrange the imports here the same way as in the other files of this package. https://github.com/golang/go/wiki/CodeReviewComments#imports

File: internal/notifier/gitea.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	"strings"
+)
+
+type Gitea struct {
+	BaseURL string
+	Token   string
+	Owner   string
+	Repo    string
+	client  *gitea.Client
+	debug   bool
```
Comment (makkes) [inline internal/notifier/gitea.go:25]: These should be public as well.

File: internal/notifier/gitea.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+var _ Interface = &Gitea{}
+
+func NewGitea(addr string, token string, certPool *x509.CertPool) (*Gitea, error) {
```
Comment (makkes) [inline internal/notifier/gitea.go:30]:
```suggestion
func NewGitea(addr, token string, certPool *x509.CertPool) (*Gitea, error) {
```

File: internal/notifier/gitea.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	if len(token) == 0 {
+		return nil, errors.New("github token cannot be empty")
```
Comment (makkes) [inline internal/notifier/gitea.go:32]:
```suggestion
		return nil, errors.New("gitea token cannot be empty")
```

File: internal/notifier/gitea.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		debug = true
+	}
+
+	return &Gitea{
+		BaseURL: host,
+		Token:   token,
+		Owner:   comp[0],
+		Repo:    comp[1],
+		client:  client,
+		debug:   debug,
```
Comment (makkes) [inline internal/notifier/gitea.go:74]:
```suggestion
	return &Gitea{
		BaseURL: host,
		Token:   token,
		Owner:   comp[0],
		Repo:    comp[1],
		client:  client,
		debug:   os.Getenv("NOTIFIER_GITEA_DEBUG") == "true",
```

File: internal/notifier/gitea.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+	host, id, err := parseGitAddress(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := url.Parse(host); err != nil {
+		return nil, err
+	}
+
+	comp := strings.Split(id, "/")
```
Comment (makkes) [inline internal/notifier/gitea.go:44]:
```suggestion
	idComponents := strings.Split(id, "/")
```

File: internal/notifier/gitea.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+	listStatusesOpts := gitea.ListStatusesOption{
+		ListOptions: gitea.ListOptions{
+			Page:     0,
+			PageSize: 50,
+		},
+	}
+	statuses, _, err := g.client.ListStatuses(g.Owner, g.Repo, rev, listStatusesOpts)
+	if err != nil {
+		return fmt.Errorf("could not list commit statuses: %v", err)
```
Comment (makkes) [inline internal/notifier/gitea.go:108]:
```suggestion
		return fmt.Errorf("could not list commit statuses: %w", err)
```

File: internal/notifier/gitea_test.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2020 The Flux authors
```
Comment (makkes) [inline internal/notifier/gitea_test.go:2]:
```suggestion
Copyright 2022 The Flux authors
```

File: internal/notifier/gitea_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
+/*
```
Comment (makkes) [inline internal/notifier/gitea_test.go:1]: Please also test the `Post` method appropriately.

**Review by makkes — CHANGES_REQUESTED** [review]

**Comment by ttys3** [comment]
the tests failed.  it complains about `config/crd/bases/notification.toolkit.fluxcd.io_providers.yaml` which I added `gitea`

**Comment by kingdonb** [comment]
I have an idea @ttys3 – it looks like you need to run `make generate` and commit the result. The error is a "dirty working tree" which just means that something in the e2e job runs `make generate` and you're supposed to already have run this.

I tried that locally but it only came up with half of the changes that failing test mentioned, the other one in `go.sum`

I guess that `make test` runs all of that

I found both the change in `go.sum` and in `notification[..]providers.yaml` like you mentioned. If you run the tests locally, you should get those changes in your working tree then you can commit them and e2e should hopefully pass after that 🎉

**Comment by ttys3** [comment]
@kingdonb `make test` will remove `gitea` from config/crd/bases/notification.toolkit.fluxcd.io_providers.yaml

`make generate` is right.

I need add `gitea` to api/v1beta1/provider_types.go, the yaml is generated from the annotation

**Comment by ttys3** [comment]
OK. fixed

PTAL

**Review by makkes — APPROVED** [review]

**Comment by makkes** [comment]
@ttys3 please squash the commits into one.

**Comment by ttys3** [comment]
squash done.

**Review by makkes — APPROVED** [review]

**Review by makkes — DISMISSED** [review]
@stefanprodan actually reminded me that we need to update the docs accordingly here. @ttys3 please add documentation about this new type to [the provider documentation page](https://github.com/fluxcd/notification-controller/blob/main/docs/spec/v1beta1/provider.md).

**Comment by ttys3** [comment]
@makkes added. PTAL

**Comment by stefanprodan** [comment]
@ttys3 this need to be ported to the v1beta2 API, we've deprecated v1beta1 in the main branch and that version is now freezed.

File: api/v1beta1/provider_types.go [hunk 10]
```diff
@@ -30,7 +30,7 @@ const (
 // ProviderSpec defines the desired state of Provider
 type ProviderSpec struct {
 	// Type of provider
-	// +kubebuilder:validation:Enum=slack;discord;msteams;rocket;generic;generic-hmac;github;gitlab;bitbucket;azuredevops;googlechat;webex;sentry;azureeventhub;telegram;lark;matrix;opsgenie;alertmanager;grafana;githubdispatch;
+	// +kubebuilder:validation:Enum=slack;discord;msteams;rocket;generic;generic-hmac;github;gitlab;gitea;bitbucket;azuredevops;googlechat;webex;sentry;azureeventhub;telegram;lark;matrix;opsgenie;alertmanager;grafana;githubdispatch;
```
Comment (stefanprodan) [inline api/v1beta1/provider_types.go:33]: Please undo all the changes to v1beta1 and move them to v1beta2, then run codegen for the API docs.

File: docs/spec/v1beta1/provider.md [hunk 11]
```diff
@@ -9,7 +9,7 @@ Spec:
 ```go
 type ProviderSpec struct {
 	// Type of provider
-	// +kubebuilder:validation:Enum=slack;discord;msteams;rocket;generic;generic-hmac;github;gitlab;bitbucket;azuredevops;googlechat;webex;sentry;azureeventhub;telegram;lark;matrix;opsgenie;githubdispatch
+	// +kubebuilder:validation:Enum=slack;discord;msteams;rocket;generic;generic-hmac;github;gitlab;gitea;bitbucket;azuredevops;googlechat;webex;sentry;azureeventhub;telegram;lark;matrix;opsgenie;githubdispatch
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:12]: Please undo this change and add Gitea to the v1beta2 docs including a section about auth, like this https://github.com/fluxcd/notification-controller/blob/main/docs/spec/v1beta2/providers.md#github

**Comment by ttys3** [comment]
CI failed. force pushed.
PTAL

**Review by stefanprodan — APPROVED** [review]
LGTM 

Thanks @ttys3 🥇

**Review by makkes — APPROVED** [review]
Great job 🎖️

**Pull request merged**